### PR TITLE
feat: improve CI, use matrix for integration tests

### PR DESCRIPTION
**Pull request by aulneau**
> Try out this version of the Stacks Wallet - download [extension builds](https://github.com/blockstack/ux/actions/runs/1026684439).

This PR does a few things:
- updates and cleans up our github workflows
- added way more caching to our yarn installs, playwright deps, and jest cache
- splits our integration tests into a matrix
- improves the speed of our tests by using [`@swc/node`](https://github.com/Brooooooklyn/swc-node/tree/master/packages/jest) 
- Unit tests got the most benefit:

| before | after |
| --- | --- |
| Test Suites: 18 passed, 18 total | Test Suites: 18 passed, 18 total |
| Tests:       42 passed, 42 total | Tests:       42 passed, 42 total |
| Snapshots:   0 total | Snapshots:   0 total | 
| Time:        72.745 s | Time:        29.08 s |
| Ran all test suites. | Ran all test suites. |
| Done in 74.15s. | Done in 29.59s.|

- Integration tests didn't benefit as much

|before | after| 
| --- | --- |
| Done in 156.69s. | Done in 145.84s.|

cc/ @aulneau @kyranjamie @fbwoolf

### CODE REVIEW

File: tests/integration/transactions/transactions.spec.ts [hunk 1]
```diff
@@ -45,15 +46,15 @@ describe(`Transactions integration tests`, () => {
...
-    const [popup] = await Promise.all([
-      browser.context.waitForEvent('page'),
-      demo.page.click(createTestSelector(TransactionsSelectors.BtnContractCall)),
-    ]);
-    const errorMsg = await popup.textContent(
-      createTestSelector(TransactionsSelectors.TransactionErrorMessage)
-    );
-    expect(errorMsg).toBeTruthy();
-  });
+  //
```
Comment (fbwoolf) [inline tests/integration/transactions/transactions.spec.ts]: Is this not working?

File: tests/integration/transactions/transactions.spec.ts [hunk 2]
```diff
@@ -1,8 +1,9 @@
-import { BrowserDriver, createTestSelector, setupBrowser } from './utils';
-import { WalletPage } from './page-objects/wallet.page';
+import { BrowserDriver, createTestSelector, setupBrowser } from '../utils';
+import { WalletPage } from '../../page-objects/wallet.page';
 import { ScreenPaths } from '@common/types';
-import { DemoPage } from './page-objects/demo.page';
-import { TransactionsSelectors } from './transactions.selectors';
+import { DemoPage } from '../../page-objects/demo.page';
+import { TransactionsSelectors } from '@tests/integration/transactions.selectors';
+// import { TransactionsSelectors } from './transactions.selectors';
```
Comment (fbwoolf) [inline tests/integration/transactions/transactions.spec.ts]: ?

**Review by fbwoolf — APPROVED** [review]
Just had a question about a test that was commented out.

File: tests/global-teardown.ts [hunk 3]
```diff
@@ -0,0 +1,5 @@
+const { teardown: teardownDevServer } = require('jest-dev-server');
+
+module.exports = async function globalTeardown() {
+  await teardownDevServer();
+};
```
Comment (kyranjamie) [inline tests/global-teardown.ts]: as this is .ts, why do we ened to use commonjs exports?

File: tests/integration/transactions/transactions.spec.ts (hunk 1 above)
Comment (aulneau) [reply tests/integration/transactions/transactions.spec.ts]: Due to the reset yesterday of testnet it was failing

File: .github/workflows/code-checks.yml [hunk 4]
```diff
@@ -51,30 +107,79 @@ jobs:
...
-      - uses: actions/checkout@v2
-      - uses: ./.github/actions/provision
+      - name: Check out Git repository
+        uses: actions/checkout@v1
+
+      - uses: actions/cache@v2
+        id: cache-node-modules
+        with:
+          path: '**/node_modules'
+          key: ${{ runner.os }}-modules-${{ hashFiles('**/yarn.lock') }}-${{ hashFiles('**/package.json') }}
```
Comment (beguene) [inline .github/workflows/code-checks.yml:174]: Nice cache key !

**Review by beguene — APPROVED** [review]
Looks good ! I like the speed up !

**Comment by aulneau** [comment]
@CharlieC3 I'd be curious your take on this PR -- it splits things out pretty much entirely, so each workflow really only has one or a couple tasks to accomplish. Additionally we're now making use of the matrix feature to split out our integration tests and run them in parallel

File: tests/integration/transactions/transactions.spec.ts (hunk 1 above)
Comment (aulneau) [reply tests/integration/transactions/transactions.spec.ts]: seems to be working now!

File: tests/global-teardown.ts (hunk 3 above)
Comment (aulneau) [reply tests/global-teardown.ts]: resolved

**Review by CharlieC3 — APPROVED** [review]
:shipit:

**Comment by CharlieC3** [comment]
> @CharlieC3 I'd be curious your take on this PR -- it splits things out pretty much entirely, so each workflow really only has one or a couple tasks to accomplish. Additionally we're now making use of the matrix feature to split out our integration tests and run them in parallel

@aulneau  It's clever to split it up by directory! I did something similar in the stacks-blockchain tests in a fork to cut the testing time in half. The downside is that it basically sucks up a good chunk of the GH Action runners for the org, especially since we're only on Free. Let's see how it goes -- If it becomes too noticeable in the following weeks by forcing other projects to wait for available runners, we might have to come up with a new plan.

**Pull request merged**